### DISCUSSION TITLE
(🚀 New Script) 🔎 ImgInspector: Quickly preview images at full size without leaving Freeplane

### DISCUSSION THREAD

**Discussion by i-plasm** [freeplane/freeplane#1948] [Show and tell]
## 🔎 ImgInspector: Quickly preview images  at full size without leaving Freeplane

***Note: the script can be installed either as an 'init' or a normal script. Installation instructions at the end.***

Hello dear community,

I'm introducing [*ImgInspector*](https://github.com/i-plasm/freeplane-scripts/wiki/ImgInspector-Updates), an automatically launched tool for full-size preview of node images.

When we insert tons of images below nodes, space quickly becomes an issue. The usual tactic for saving mindmap space is diminishing image zoom, but images become hard to distingish if set too low. *ImgInspector* intends to solve these issues.

![demo_ImgInspector](https://github.com/user-attachments/assets/7d8a6752-7f91-413a-983a-a48ae8d67e31)

### Features

As seen, *ImgInspector* is an automatically launched popup, triggered when the user hovers over a node's image area. Some of its features are:

- Full-size image preview on mouse hover
- Full-size image preview on external frame
- Copy full image URL
- Automatically launches and closes itself

### ImgInspector and Sketcher-Intelliflow as companion scripts

The scripts [***Sketcher (Intelliflow)***](https://github.com/freeplane/freeplane/discussions/1496) and ***ImgInspector*** complement each other - the first one deals with image creation and edition, the other with image previewing. 

With the **Sketcher Templates** feature, using tools like [*draw.io*](https://github.com/jgraph/drawio-desktop) for embedding and editing all sorts of diagrams becomes really easy. Diagrams tend to have large dimensions. *ImgInspector* allows a map featuring many images of large dimensions to save tons of space while offering easy full-size previewing. 

## Permissions

All four script permissions are required (set them in `Tools -> Preferences -> Plug-ins`): 1) Permit file read, 2) write, 3) network, and 4) execute other applications. **IMPORTANT! This is not ideal, so before granting permissions, please make sure the scripts you have installed are safe. In the case of ImgInspector, if permissions aren't granted, an "access denied java.awt.AWTPermission" exception will be thrown. As a workaround, I might convert the script to an "Add-On" so that the permissions are localized to it, without the need to activate permissions globally.**

## Installation and Launching

You may install the script as a regular or an 'init' script (scripts that are executed on startup). Steps:

1. Download the groovy script from the [latest ImgInspector Updates](https://github.com/i-plasm/freeplane-scripts/wiki/ImgInspector-Updates)
1. Open the user directory (`Tools->Open user directory`)
1. Copy the groovy file to the `/scripts/init/` subfolder if you want to install as an init script, or to the `/scripts/` subfolder for a regular script install.
1. Restart Freeplane
1. If installed as an init script, you don't need to launch it. If installed as a regular script, you need to run it once at: `i-plasm -> Misc Utils -> Freeplane Img Inspector`

---

Note: SVG and TIFF formats are not supported yet.

I hope you will find it helpful!

**Comment by Illegit** [comment]
Hi I-plasm, your tool is very useful. 
For my personal uses, I need to have the image bigger than the full size. E.g.: if I have a zoom of 121% the image is shown at 100% and the same if my zoom is 100% or 56%. My question is: may I get an image that is a % bigger than my actual zoom %, and where I set it in your script? Or better to have this option in the pop-up menu?
Thanks & compliments.

**Comment by euu2021** [comment]
Brilliant! Thanks!

An issue that I see is that, if the FP window is on my secondary monitor, the image popup opens on the primary monitor, and I need to make the mouse hover on the primary window in order to make the image popup disappear.

**Comment by i-plasm** [comment]
Hi  @euu2021 and @Illegit ,

Thank you for your feedback.

(Re: the dual monitor set-up issue) @euu2021 I fixed some bugs in the latest version (https://github.com/i-plasm/freeplane-scripts/wiki/ImgInspector-Updates). I've tested using two monitors and Windows OS. Please let me know if it works as expected. By the way, you can also use `ESC` on the preview windows to close them.

(Re: image zooming) @Illegit  Can you please clarify: Are you referring to zooming in/out manually from inside the preview popup? Or do you mean reading/remembering a persistent zoom value for each node image and previewing it with such value?

**Reply by euu2021** [reply]
It's working perfectly. Thanks.

**Comment by Illegit** [comment]
Hi I-plasm, I meant the 2nd (a persistent zooming value) because I thought
the 1st (zooming in out in the preview) was difficult to activate, but I
prefer this last one (zooming in out in the preview) if it is easy to
achieve... I like your tool.

Il lun 5 ago 2024, 15:36 bbarbosa ***@***.***> ha scritto:

**Reply by i-plasm** [reply]
Thanks, I'll consider it.

**Comment by acebone** [comment]
This is genious!

I am on linux (ubuntu 23.10) with dual monitor. 

I get the preview on the primary monitor even if freeplane is on the secondary. It's not a biggie, because I can just hit escape to close it.

The little popup appears where it's supposed to. I doesn't draw the magnifier-glass, but hovering over where it should've been drawn still works.

![image](https://github.com/user-attachments/assets/74fe61e0-e5a5-465c-8a28-4d46e2b7aa7c)

These are tiny errors in my workflow, and doesn't really matter to me. But I thought I'd let you know.

Thanks for a great idea and a great script!

**Reply by i-plasm** [reply]
Hello @acebone . Thank you for your feedback. Regarding the dual monitor issue, @euu2021  reported a bug (https://github.com/freeplane/freeplane/discussions/1948#discussioncomment-10234861) with the exact same description a few weeks ago. I fixed it, so I am wondering if your script version is outdated. Please try the [latest update](https://github.com/i-plasm/freeplane-scripts/wiki/ImgInspector-Updates). Make sure when you try to download the latest version from your browser that it is not displaying some old cached version. You may need to perform a "hard refresh/reload" of the script web page to clear the webpage cache. Please let me know if the issue persists. 

Regarding the icons issue, unfortunately Java has well known problems with dealing with unicode characters - that's why you see strange things like a square, etc. I also observed problems on macos, but on windows it seems to display fine. I might employ a workaround in the future that avoids using unicode.

**Reply by acebone** [reply]
It was indeed old, so I just dl'ed and tested  "Last Update: 2024-09-18"

And it also shows the preview on the primary screen, even if FP is on the secondary.

Note that I use Ubuntu 23.10 - that might explain why a fix that works on Windows doesn't work on my system.

I'd be happy to test for you, if that could help, but I can easily live with it as it is.

**Reply by i-plasm** [reply]
> I'd be happy to test for you, if that could help

Thanks for that. I finally tested on Linux mint (XFCE) and I reproduced the bug you describe, and also I noticed geometry problems on large images. Because of that and because of the invisible icon issue, the user experience is bad. None of those issues occur on Windows so your report has been helpful. I will post a fix soon.

**Reply by acebone** [reply]
Great!

Meanwhile I've got another thing that you might want to look at:

Your script overrides the default right-click menu, so you can't do remove, change, open in viewer or reset zoom anymore.

I really miss being able to do those - especially change and open in viewer.

**Reply by i-plasm** [reply]
I just published a new [update](https://github.com/i-plasm/freeplane-scripts/wiki/ImgInspector-Updates). I fixed several glitches and misbehaviors on non-Windows OS, and also implemented the use of alternative icons for non-Windows OS. Please let me know if you try it out. 

> Your script overrides the default right-click menu, so you can't do remove, change, open in viewer or reset zoom anymore.

The script hasn't in any way blocked or overridden the right-click menu on Freeplane's image component. I can launch it without issues on all OS I've tried. I'm not sure if that's what you meant.
